source/common/grpc/common.cc

```diff
@@ -269,5 +286,98 @@ std::string Common::typeUrl(const std::string& qualified_name) {
   return typeUrlPrefix() + "/" + qualified_name;
 }
 
+struct BufferInstanceContainer {
+  BufferInstanceContainer(int ref_count, Buffer::InstancePtr buffer)
+      : ref_count_(ref_count), buffer_(std::move(buffer)) {}
+  std::atomic<int> ref_count_;
+  Buffer::InstancePtr buffer_;
+};
+
+static void derefBufferInstanceContainer(void* container_ptr) {
+  auto container = reinterpret_cast<BufferInstanceContainer*>(container_ptr);
+  container->ref_count_--;
+  if (container->ref_count_ <= 0) {
+    delete container;
+  }
+}
+
+grpc::ByteBuffer Common::makeByteBuffer(Buffer::InstancePtr bufferInstance) {
+  if (!bufferInstance) {
+    return {};
+  }
+  Buffer::RawSlice oneRawSlice;
+  // NB: we need to pass in >= 1 in order to get the real "n" (see Buffer::Instance for details).
+  int nSlices = bufferInstance->getRawSlices(&oneRawSlice, 1);
+  if (nSlices <= 0) {
+    return {};
+  }
+  auto container = new BufferInstanceContainer{nSlices, std::move(bufferInstance)};
+  if (nSlices == 1) {
+    grpc::Slice oneSlice(oneRawSlice.mem_, oneRawSlice.len_, &derefBufferInstanceContainer,
+                         container);
+    return {&oneSlice, 1};
+  }
+  STACK_ARRAY(manyRawSlices, Buffer::RawSlice, nSlices);
+  container->buffer_->getRawSlices(manyRawSlices.begin(), nSlices);
+  std::vector<grpc::Slice> slices;
+  slices.reserve(nSlices);
+  for (int i = 0; i < nSlices; i++) {
+    slices.emplace_back(manyRawSlices[i].mem_, manyRawSlices[i].len_, &derefBufferInstanceContainer,
+                        container);
+  }
+  return {&slices[0], slices.size()};
+}
+
+struct ByteBufferContainer {
+  ByteBufferContainer(int ref_count) : ref_count_(ref_count) {}
+  ~ByteBufferContainer() { ::free(fragments); }
+  std::atomic<int> ref_count_;
```

Comment:
I don't think this one needs to be atomic, Envoy won't release cross-thread.

Reply:
Once you can interconvert these then it isn't clear who is actually doing the free, gRPC or the standard envoy layer.  It seems safer to me to do them both atomic to account for unanticipated use cases.  Follow up if you disagree and I'll change it.

Reply:
Removed the atomic on the ByteBufferContainer as Buffer::Instance is not thread safe.  Left the other as grpc::ByteBuffer is thread safe AFAICT.